gateway/api/services.py

```diff
@@ -65,3 +75,135 @@ def save(serializer, author, artifact) -> Program:
 
         return program
+
+    @staticmethod
+    def find_one_by_title(title, author) -> Program:
+        """
+        It returns the last created Program by title from an author
+
+        Args:
+            title: program title
+            author: user tipically got it from request
+
+        Returns:
+            program: Program instance found
+        """
+
+        logger.debug("Filtering Program by title[%s] and author [%s]", title, author)
+        program = (
+            Program.objects.filter(title=title, author=author)
+            .order_by("-created")
+            .first()
+        )
+
+        if program is None:
+            logger.error("Program [%s] by author [%s] not found", title, author)
+            raise ResourceNotFoundException("Program [{title}] was not found")
+
+        return program
+
+
+class JobConfigService:
+    """
+    JobConfig service allocate the logic related with job configuration
+    """
+
+    @staticmethod
+    def save_with_serializer(serializer) -> JobConfig:
+        """
+        It returns a new JobConfig from its serializer
+
+        Args:
+            serializer: JobConfig serializer from the model
+
+        Returns:
+            JobConfig: new JobConfig instance
+        """
+
+        # It would be nice if we could unify all the saves logic in one unique entry-point
+        try:
+            jobconfig = serializer.save()
+        except (Exception) as save_job_config_exception:
+            logger.error(
+                "Exception was caught saving a JobConfig. \n Error trace: %s",
+                save_job_config_exception,
+            )
+            raise InternalServerErrorException(
+                "Unexpected error saving the configuration of the job"
+            ) from save_job_config_exception
+
+        logger.debug("JobConfig [%s] saved", jobconfig.id)
+
+        return jobconfig
+
+
+class JobService:
+    """
+    Job service allocate the logic related with a job
+    """
+
+    @staticmethod
+    def save(
+        program: Program,
+        arguments,
+        author,
+        jobconfig: JobConfig,
+        token: str,
+        carrier,
+        status=Job.QUEUED,
+    ) -> Job:
+        """
+        Creates or updates a job
+
+        Args:
+            program: instance of a Program
+            arguments: arguments from the serializer
+            author: author from the request
+            jobconfig: instance of a JobConfig
+            token: token from the request after being decoded
+            carrier: object injected from TraceContextTextMapPropagator
+            status: status of the job, QUEUED by default
+
+        Returns:
+            Job instance
+        """
+
+        job = None
+        try:
+            job = Job(
+                program=program,
+                arguments=arguments,
+                author=author,
+                status=status,
+                config=jobconfig,
+            )
+            job.save()
```

Comment:
This is not related to this refactoring but this `save()` is not good.  The job is created here and updated with the environment later may cause an issue.  The scheduler may find the job between these 2 save() and execute without the environment information.

Reply:
Mmm... You are totally right. I tried to use one unique `save` but I couldn't figure out how. Any ideas @IceKhan13 ?

Reply:
Two options that come to my mind:
- Have a new state before `QUEUED` that reflects some kind of an `under state preparation`.
- Remove job.id from [environment variables](https://github.com/Qiskit-Extensions/quantum-serverless/pull/1147/files#diff-b1f12a07f91db92d539b60433b3f9831281e139afeee593565c3c25563f3c206L131) but I think we need to have it

Reply:
First solution could be implemented in a new pull request and the good thing is that is not going to require any migration or breaking change.